rfcs/0011-reusable-reactors-folder.md

```diff
@@ -0,0 +1,115 @@
+- **`Lingo.toml`**: Configuration file crucial for project sharing, storing Lingo configuration details including external LF projects to include (refer to [Lingo Documentation](https://github.com/lf-lang/lingo?tab=readme-ov-file#the-toml-based-package-configurations) for further details). The internal structure of this file will be discussed in another RFC.
+
+The refined structure will appear as follows:
+```shell
+├── root
+│   ├── bin/
+│   ├── include/
+│   ├── lib/ # Directory for storing reusable reactors
+│   │   ├── Input.lf # Ex: reactor capturing external inputs (e.g., Microphone, Camera)
+│   │   └── ComputerVision.lf # Ex: reactor performing computer vision tasks (e.g., object detection, face recognition)
+│   ├── src/
+│   ├── src-gen/
+│   ├── fed-gen/ # Only for federated programs (instead of src-gen)
+│   └── Lingo.toml # Configuration file for Lingo Package Manager
+```
+The `lib/` directory is intended to store only reusable reactors, allowing them to be shared across multiple projects and reducing the need for reimplementation. The `src/` folder, on the other hand, is for executable programs that might use reactors defined in the `lib/` folder to develop a program. Generally, `src/` should contain only files that the developer compiles and executes. Note that the content of directories such as `src-gen`, `fed-gen`, and `include` are outside the scope of this RFC.
+
+This structured approach promotes efficient code reuse, simplifies integration of external resources, and fosters collaborative development practices within the LF community.
+
+### Supporting reactors
+[supporting-reactors]: #supporting-reactors
+
+According to this proposal, the `lib/` directory is designated for reusable reactors, while the `src/` directory is reserved for executable programs. In some projects, there may be reactors that are used internally by a library but are not intended to be exposed to its users. These can be considered "supporting reactors." To clearly mark such reactors as internal and not for public use, we propose using a `private/` directory within the `lib/` folder. This directory can exist at any level of hierarchy within `lib/` and signifies that its contents are private and should not be exposed to users.
+
+For example:
+
+```shell
+├── root
+│   ├── bin/
+│   ├── include/
+│   ├── lib/
+│   │   ├── Input.lf
+│   │   ├── ComputerVision.lf
+│   │   └── private/
+│   │       ├── SupportingReactor1.lf
+│   │       └── SupportingReactor2.lf
+│   ├── src/
+│   ├── src-gen/
+│   ├── fed-gen/  # Only for federated programs (instead of src-gen)
+│   └── Lingo.toml
+```
+
+In this structure, the `private/` directory contains `SupportingReactor1.lf` and `SupportingReactor2.lf`. These reactors are intended to be extended by other reactors or imported for specific functions, but are not meant to be exposed to users of the library. This convention helps maintain a clear separation between public and internal components of the library.
```

Comment:
I think it's more common to use a visibility modifier or export statement for stuff like this. What's the reason for doing this in the file system?

Reply:
Presumably, the user we're trying to hide things from is one that pulls a package from a remote repo, Since they use `lingo` for this, we could also put `lingo` in charge of the visibility issue. For instance, one could whitelist or blacklist certain files or directories in the `Lingo.toml` file. This kind of mechanism you see a lot in linters, package managers, etc.

Reply:
Lingo only drags files around. 

I would put this check into lfc, and do a very simple check:

```
if "/private/" in include_path:
  print("warning you are trying to import private files")
```

Reply:
I was thinking the proposal was to build checks into `lfc` to enforce visibility, but this appears not to be the case.